homeassistant/components/heos/config_flow.py

```diff
@@ -100,3 +116,72 @@ async def async_step_reconfigure(
             data_schema=vol.Schema({vol.Required(CONF_HOST, default=host): str}),
             errors=errors,
         )
+
+
+OPTIONS_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_USERNAME): selector.TextSelector(),
+        vol.Optional(CONF_PASSWORD): selector.TextSelector(
+            selector.TextSelectorConfig(type=selector.TextSelectorType.PASSWORD)
+        ),
+    }
+)
+
+
+class HeosOptionsFlowHandler(OptionsFlow):
+    """Define HEOS options flow."""
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Manage the options."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            authentication = CONF_USERNAME in user_input or CONF_PASSWORD in user_input
+            if authentication and CONF_USERNAME not in user_input:
+                errors[CONF_USERNAME] = "username_missing"
+            if authentication and CONF_PASSWORD not in user_input:
+                errors[CONF_PASSWORD] = "password_missing"
```

Comment:
I think we can avoid doing this using `vol.All` in the schema, I'm just not sure if we have a prebuilt validator for this

Reply:
How do I use `vol.All` this way? I searched the code base and didn't see other examples.

Reply:
`zwave_js` has a couple of validators check `has_at_least_one_node` https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/services.py#L263

Reply:
Using `vol.All` inside `vol.Schema` or at the top level causes a `ValueError: dictionary update sequence element #0 has length 3; 2 is required` at https://github.com/home-assistant/core/blob/adb1fbbbc453bf6c0e73fe429396970274ccdcf6/homeassistant/helpers/data_entry_flow.py#L52

Here's what I have (that does not work):

```python
def has_both_or_neither(val: dict[str, Any]) -> dict[str, Any]:
    """Validate that at least one node is specified."""
    if val.get(CONF_USERNAME) ^ val.get(CONF_PASSWORD):
        raise vol.Invalid(
            "Either both or neither of username and password must be specified"
        )
    return val


OPTIONS_SCHEMA = vol.Schema(
    vol.All(
        {
            vol.Optional(CONF_USERNAME): selector.TextSelector(),
            vol.Optional(CONF_PASSWORD): selector.TextSelector(
                selector.TextSelectorConfig(type=selector.TextSelectorType.PASSWORD)
            ),
        },
        has_both_or_neither,
    )
)
```

I couldn't find any other options flow schemas that use `vol.All` this way nor use custom validation functions. FWIW, I followed the same pattern that the OpenSky integration did: https://github.com/home-assistant/core/blob/adb1fbbbc453bf6c0e73fe429396970274ccdcf6/homeassistant/components/opensky/config_flow.py#L88-L131

What do you suggest?

Reply:
Nice, that works. One drawback is that it will show an unlocalized error message on the front-end. In the form, it will show:
> some but not all values in the same group of inclusion 'auth' @ data[<auth>]

I added the `msg` parameter to show something more user friendly, however, it's still unlocalized.
> Username and Password must be supplied together

Is there a way to return a localized string here?

Reply:
no idea tbh

Reply:
In Opensky I opted to just do it manually so I can give a nice translated value
https://github.com/home-assistant/core/blob/c54ed53a818728807786f52c8eb789da445ed8db/homeassistant/components/opensky/config_flow.py#L94-L102

Reply:
Lol, this is where we started 😅

Reply:
Is it beautiful code? Maybe not. Does it do the job and is it user friendly? Yes. So yea

Reply:
I'll revert back so we get the localized error message. 😃

Reply:
@zweckj reset from the last change. Ready for final review!